zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+
+    /// Get a mutable reference to the node at `index`.
+    fn get_node_mut(&mut self, index: u32) -> Result<&mut Node<T>, DispatchError>;
+
+    /// Get the node which belongs to `account`.
+    fn map_account_to_index(&self, account: &T::AccountId) -> Result<u32, DispatchError>;
+
+    /// Mutate the node at `index` using `mutator`.
+    fn mutate_node<F>(&mut self, index: u32, mutator: F) -> DispatchResult
+    where
+        F: FnOnce(&mut Node<T>) -> DispatchResult;
+
+    /// Return the maximum allowed depth of the tree.
+    fn max_depth() -> u32;
+
+    /// Return the maximum allowed amount of nodes in the tree.
+    fn max_node_count() -> u32;
+}
+
+impl<T, U> LiquidityTreeHelper<T> for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn propagate_fees_to_node(&mut self, index: u32) -> DispatchResult {
+        let path = self.path_to_node(index)?;
+        for i in path {
+            self.propagate_fees(i)?;
+        }
+        Ok(())
+    }
+
+    fn propagate_fees(&mut self, index: u32) -> DispatchResult {
+        let node = self.get_node(index)?;
+        if node.total_stake()? == Zero::zero() {
+            return Ok(()); // Don't propagate if there are no LPs under this node.
+        }
+        // Temporary storage to ensure that the borrow checker doesn't get upset.
+        let descendant_stake = node.descendant_stake;
+        if node.is_leaf() {
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&node.lazy_fees)?;
+                Ok(())
+            })?;
+        } else {
+            // The lazy fees that will be propagated down the tree.
+            let mut remaining_lazy_fees =
+                node.descendant_stake.bmul_bdiv(node.lazy_fees, node.total_stake()?)?;
+            // The fees that stay at this node.
+            let fees = node.lazy_fees.checked_sub_res(&remaining_lazy_fees)?;
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&fees)?;
+                Ok(())
+            })?;
+            let child_indices = self.children(index)?;
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(0).unwrap_or(&None) {
+                Some(lhs_index) => {
+                    self.mutate_node(*lhs_index, |lhs_node| {
+                        let child_lazy_fees = lhs_node
+                            .total_stake()?
+                            .bmul_bdiv(remaining_lazy_fees, descendant_stake)?;
+                        lhs_node.lazy_fees =
+                            lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                        remaining_lazy_fees =
+                            remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
```

Comment:
Because you can't iterate over a tuple.

Reply:
Because of your comment there I was reminded about using a tuple.

Reply:
135c15a92d83d102075e93feba558397232d4205

I think this is the right way to go. Tuples are kinda off because you can't iterate over them, but arrays don't seem ideal either given our no panickers policy.

Reply:
135c15a92d83d102075e93feba558397232d4205